### DISCUSSION TITLE
Reports and email

### DISCUSSION THREAD

**Discussion by nomad-cyanide** [netbox-community/netbox#13708] [General]
Hi

After one of my reports have run, I need to send an email to my boss with the results, but currently I am just sending the self._results to myself and that is just CSV data, which I know my boss can't/won't read. 
Are there other clever functions relating to formatting the output or do I need to process the self._results somehow, before putting it the email-body?

Best regards

**Comment by mtinberg** [comment]
I don't think there is anything special to email from Reports, I'm just using f-strings to format a message

def send_email(results, report_name, subject, header, ***@***.***"):

    # Early return if there are no failures
    if results._results[report_name]['failure'] == 0:
        return
    # Process the result log data into sections
    logs = {
        'failure': "",
        'warning': "",
        'info': "",
        'success': "",
        'default': ""
    }
    for log in results._results[report_name]['log']:
        if log[1] in ['failure', 'warning', 'info', 'success']:
            logs[log[1]] += f"{log[2]} : {log[4]}\n"
    msg = EmailMessage()
    msg['Subject'] = subject
    msg['From'] = 'netbox'
    msg['To'] = to
    msg.set_content(f"""

{subject}
====================

{header}

Failure {results._results[report_name]['failure']}
Warning {results._results[report_name]['warning']}
Info {results._results[report_name]['info']}
Success {results._results[report_name]['success']}

Logs
------

Failure
-------

{logs['failure']}

Warning
-------

{logs['warning']}

Info
----

{logs['info']}

Success
-------

{logs['success']}

""")

    mailer = smtplib.SMTP('localhost')
    mailer.send_message(msg)
    mailer.quit()


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Ian S. ***@***.***>
Sent: Thursday, September 7, 2023 4:08 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Reports and email (Discussion #13708)


Hi

After one of my reports have run, I need to send an email to my boss with the results, but currently I am just sending the self._results to myself and that is just CSV data, which I know my boss can't/won't read.
Are there other clever functions relating to formatting the output or do I need to process the self._results somehow, before putting it the email-body?

Best regards

**Reply by nomad-cyanide** [reply]
Thanks, I was trying to do something like that, but the report just comes up with "error", with no apparent way of seeing said error. Kinda frustrating.

**Comment by nomad-cyanide** [comment]
I could not get solution from Mr. Mark Tinberg to work either, so I have tried doing something simpler.

When I issue this post_run command I get the CSV data in an email, in the form of a dictionary.

> send_mail('Config backup status report','Config backup status:\n%s'%(self._results),'noreply-netbox@test-site',['xxx@yyy.zzz'])

I copied that dict into a variable called report_result and tested the code in python3 to where the output was as I wanted, using print, instead of send_mail. 
Now I have implemented that code in the report, but the report still fails with "Errored" and no emails are sent. :( 

Code for post_run is like this:
```   
 def post_run(self):
        mail_body = ""
        error_line = ""
        report_result = self._results
        backup_log = report_result["test_backup_status"]["log"]
        for log_entry in backup_log:
            warning_type = log_entry[1]
            device_name = log_entry[2]
            device_link = log_entry[3]
            error_msg =  log_entry[4]
            error_line = warning_type+": "+device_name+"- Msg: "+error_msg+" - https://netbox-test"+device_link+"\n"
            mail_body += error_line
        send_mail('Config backup status report','Config backup status:\n%s'%(mail_body),'noreply-netbox@test-site',['xxx@yyy.zzz])
```

journalctl -xr just shows "Error during execution of report". Nothing useful.

How do I see what is going wrong, since the code is working in "normal" python?

**Reply by nomad-cyanide** [reply]
Figured it out myself. 
A print message puts output into journal log, that can be viewed with journalctl -xr command.

**Comment by zmaster7** [comment]
Not exactly with Reports, but I was able to use jinja2 in a custom script with a HTML template to email the results as HTML, which can be printed to PDF if need be. 

The relevant bits of code would look similar to:

```
from datetime import datetime
from email.message import EmailMessage
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.utils import formatdate
from jinja2 import Environment, FileSystemLoader
import os
import smtplib

timestamp = datetime.now()
filepath = os.path.dirname(os.path.abspath(__file__))
templateDir = os.path.join(filePath, 'Templates')
env = Environment( loader = FileSystemLoader(templateDir) )
template = env.get_template('Template.html')
templateRender = template.render(templateResults = results, timestamp = timestamp.strftime("%m/%d/%Y %H:%M:%S"))
templateName = f'Report.html'

if data['emailAddress']:
	emailFrom = "email@somewhere.org"
	emailTo = data['emailAddress']
	emailSubject = f'Netbox - Report'
	emailBody = MIMEText(f'<html><body>Report Run By: {self.request.user.username}<br>Time Run: {timestamp.strftime("%m/%d/%Y %H:%M:%S")}</body></html>', 'html')
	emailServer = "mailserver.somewhere.org"
	emailPort = 25
	
	msg = MIMEMultipart()
	
	msg['From'] = emailFrom
	msg['To'] = emailTo
	msg['Date'] = formatdate(localtime=True)
	msg['Subject'] = emailSubject
	msg.attach(emailBody)
	
	attachment = MIMEText(templateRender, 'html')
	attachment.add_header('Content-Disposition', 'attachment', filename=templateName)
	msg.attach(attachment)
	
	try:
		server = smtplib.SMTP(emailServer,emailPort)
		server.starttls()
		emailText = msg.as_string()
		server.sendmail(emailFrom, emailTo, emailText)
		server.quit()
		
		self.log_info(f'Email Sent To: [{emailTo}]. It may take a few minutes to appear in your inbox.') 
	except:
		self.log_warning(f'Email failed to send to: [{emailTo}].') 
else:
	self.log_info(f'Email not specified, HTML report not sent.')
```

and a very basic jinja2 HTML template would look like this:

```
<html>
<head>
	<title>Netbox - Report</title>
	<style>
	</style>
</head>
<body>
<h1>Netbox - Report</h1>

{%- if timestamp -%}
<h2>Report Generated: {{ timestamp }}</h4>
{%- endif -%}

{%- if templateResults -%}
<table class="Results">
	<tr>
		<th>Header1</th>
		<th>Header2</th>
		<th>Header3</th>
	</tr>
	{%- for result in templateResults -%}
	<tr>
		<td>{{ result['Header1'] }}</td>
		<td>{{ result['Header2'] }}</td>
		<td>{{ result['Header3'] }}</td>
	</tr>
	{%- endfor -%}
</table>
{%- endif -%}
</body>
</html>
```
